alexandria/core/views.py

```diff
@@ -64,6 +64,27 @@ class DocumentViewSet(PermissionViewMixin, VisibilityViewMixin, views.ModelViewS
     filterset_class = DocumentFilterSet
     search_fields = ("title", "files__name", "tags__name", "description")
 
+    def update(self, request, *args, **kwargs):
+        instance = self.get_object()
+        tags_before_patch = instance.tags.all()
+        tags_after_patch = request.data.get("tags")
+
+        # check if we're removing tags
+        for tag in tags_before_patch:
+            # tags_after_patch is a list of dicts
+            tag_as_dict = {"id": tag.slug, "type": "tags"}
+            if tags_after_patch is None or tag_as_dict not in tags_after_patch:
+                # check if the tag to be deleted is used in other documents
+                # if not delete the tag completely
+                if (
+                    not models.Document.objects.filter(tags=tag)
+                    .exclude(id=instance.id)
+                    .exists()
+                ):
+                    models.Tag.objects.get(slug=tag.slug).delete()
+
+        return super().update(request, *args, **kwargs)
```

Comment:
This is quite complicated, and runs a lot of SQL for the expected result. You could just (irregardless of the actual change in the document) do this:

```python
Tag.objects.all().filter(documents__pk__isnull=True).delete()
```

Reply:
It does delete all currently unused tags, and I have not thought about that.
The only issue I see with your solution is that it doesn't delete tags that are going to be removed by the new request. Deleting a tag completely would require two requests. This could potentially lead to some confusion in frontend, as a deleted tag could still be seen for a while.

Reply:
Right - this wouldn't work correctly right now. 

There are a few options though:
* You could move the code into a `post_delete` signal handler.
* Call `super().update(...)` first, then do the cleanup before returning - This would be my preferred solution